### PR TITLE
lowdown: fix aarch64-darwin by removing deprecated sandbox

### PR DESCRIPTION
**Pull request by hlolli**
###### Motivation for this change

Currently latest nix doesn't build on aarch64 due to failure from lowdown.
The current regression test provided from lowdown falsly pass.
So I wrote one very basic install-check.

Some interesting discussions about the (bad and deprecated) sandbox.h
https://developer.apple.com/forums/thread/661939

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/typesetting/lowdown/default.nix [hunk 1]
```diff
@@ -24,12 +24,20 @@ stdenv.mkDerivation rec {
...
   # Fix lib extension so that fixDarwinDylibNames detects it
-  postInstall = lib.optionalString stdenv.isDarwin ''
+  postInstall = lib.optionalString (stdenv.isDarwin && !stdenv.isAarch64) ''
     mv $lib/lib/liblowdown.{so,dylib}
   '';
 
   patches = lib.optional (!stdenv.hostPlatform.isStatic) ./shared.patch;
 
+  patchPhase = lib.optionalString (stdenv.isDarwin && stdenv.isAarch64) ''
+    substituteInPlace main.c \
```
Comment (sternenseemann) [inline pkgs/tools/typesetting/lowdown/default.nix:34]: You should be able to set `HAVE_SANDBOX_INIT=0` in `configure.local` instead which I'd prefer as it is the intended way setting this (and doesn't require us to touch the real source code).

File: pkgs/tools/typesetting/lowdown/default.nix [hunk 2]
```diff
@@ -24,12 +24,20 @@ stdenv.mkDerivation rec {
   '';
 
   # Fix lib extension so that fixDarwinDylibNames detects it
-  postInstall = lib.optionalString stdenv.isDarwin ''
+  postInstall = lib.optionalString (stdenv.isDarwin && !stdenv.isAarch64) ''
```
Comment (sternenseemann) [inline pkgs/tools/typesetting/lowdown/default.nix:27]: Just curious: Is aarch64-darwin somehow different here or does fixDarwinDylibNames work differerntly?

**Review by sternenseemann — CHANGES_REQUESTED** [review]

File: pkgs/tools/typesetting/lowdown/default.nix (hunk 1 above)
Comment (hlolli) [reply pkgs/tools/typesetting/lowdown/default.nix:34]: I tried that, but thinking back, I probably didn't set this env-var right. Let me try if I get better luck.

File: pkgs/tools/typesetting/lowdown/default.nix (hunk 2 above)
Comment (hlolli) [reply pkgs/tools/typesetting/lowdown/default.nix:27]: This probably isn't needed when I think about it.
aarch64 is different in few ways (like needs sigtool and calls launchd's not in x86 to name a few)

File: pkgs/tools/typesetting/lowdown/default.nix (hunk 1 above)
Comment (hlolli) [reply pkgs/tools/typesetting/lowdown/default.nix]: configure.local did the trick, I'm amazed 👍

File: pkgs/tools/typesetting/lowdown/default.nix (hunk 2 above)
Comment (hlolli) [reply pkgs/tools/typesetting/lowdown/default.nix]: wasn't needed, so it's removed

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 125004` at b62962b2 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built successfully:</summary>
  <ul>
    <li>hydra-unstable</li>
    <li>lowdown</li>
    <li>nix-direnv</li>
    <li>nix-update</li>
    <li>nixUnstable</li>
    <li>nixpkgs-review</li>
  </ul>
</details>

File: pkgs/tools/typesetting/lowdown/default.nix [hunk 3]
```diff
@@ -14,6 +14,9 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [ which ]
     ++ lib.optionals stdenv.isDarwin [ fixDarwinDylibNames ];
 
+  preConfigure = lib.optionalString (stdenv.isDarwin && stdenv.isAarch64)
```
Comment (veprbl) [inline pkgs/tools/typesetting/lowdown/default.nix:17]: Please use this style:
```suggestion
  preConfigure = lib.optionalString (stdenv.isDarwin && stdenv.isAarch64) ''
    echo 'HAVE_SANDBOX_INIT=0' > configure.local
  ''
```
There is a subtle edge case with overrides where adding `""` strings together (imagine adding extra `preConfigure` steps in an override) will result a missing newline and broken bash code.

File: pkgs/tools/typesetting/lowdown/default.nix [hunk 4]
```diff
@@ -30,6 +33,9 @@ stdenv.mkDerivation rec {
 
   patches = lib.optional (!stdenv.hostPlatform.isStatic) ./shared.patch;
 
+  doInstallCheck = stdenv.hostPlatform == stdenv.buildPlatform;
+  installCheckPhase = "echo '# TEST' > test.md; $out/bin/lowdown test.md";
```
Comment (veprbl) [inline pkgs/tools/typesetting/lowdown/default.nix:37]:
```suggestion
  installCheckPhase = ''
    runHook preInstallCheck
    echo '# TEST' > test.md; $out/bin/lowdown test.md
    runHook postInstallCheck
  '';
```

**Comment by veprbl** [comment]
@GrahamcOfBorg build lowdown

**Comment by hlolli** [comment]
@veprbl you're fast! but I added there as well $out/bin to the postInstall check, it doesn't find the binary without some help it seems.

**Comment by veprbl** [comment]
@GrahamcOfBorg build lowdown

**Review by veprbl — APPROVED** [review]

**Comment by veprbl** [comment]
Result of `nixpkgs-review pr 125004` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>lowdown</li>
    <li>nix-direnv</li>
    <li>nix-review (nixpkgs-review)</li>
    <li>nix-update</li>
    <li>nixFlakes (nixUnstable)</li>
  </ul>
</details>

**Comment by andir** [comment]
This PR doesn't follow our contribution guidelines for commit messages. Please try a bit harder to follow it. If senior contributors don't do that we have zero chance to enforce it as they aren't leading with good examples.

**Comment by veprbl** [comment]
Ouch, this was intended to be squashed on merge...

**Comment by hlolli** [comment]
> This PR doesn't follow our contribution guidelines for commit messages. Please try a bit harder to follow it. If senior contributors don't do that we have zero chance to enforce it as they aren't leading with good examples.

that's very fair point!

I'm less senior than having merge rights, but I'm in the habit of squasing and renaming the commit messages.
But I see now this was probably aimed at @veprbl altough I can do better too.

**Pull request merged**